libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+	}
+	CloseHandle (th);
+	return size;
+}
+
+int w32_reg_write(RDebug *dbg, int type, const ut8 *buf, int size) {
+	DWORD flags = THREAD_SUSPEND_RESUME | THREAD_SET_CONTEXT;
+	if (dbg->bits == R_SYS_BITS_64) {
+		flags |= THREAD_QUERY_INFORMATION;
+	}
+	HANDLE th = OpenThread (flags, FALSE, dbg->tid);
+	if (th == (HANDLE)NULL) {
+		r_sys_perror ("w32_reg_write/OpenThread");
+		return false;
+	}
+	// Always suspend
+	if (suspend_thread (th, dbg->bits) == -1) {
+		CloseHandle (th);
+		return false;
+	}
+	bool ret = set_thread_context (th, buf, size, dbg->bits);
+	// Always resume
+	if (resume_thread (th, dbg->bits) == -1) {
+		ret = false;
+	}
+	CloseHandle (th);
+	return ret;
+}
+
+int w32_attach(RDebug *dbg, int pid) {
+	RIOW32Dbg *rio = dbg->user;
+	HANDLE ph = OpenProcess (PROCESS_ALL_ACCESS, FALSE, pid);
+	if (ph == (HANDLE)NULL) {
+		return -1;
+	}
+	if (!DebugActiveProcess (pid)) {
+		CloseHandle (ph);
+		return -1;
+	}
+	RList *threads = r_list_new ();
+	if (!threads) {
+		CloseHandle (ph);
+		return -1;
+	}
+	threads = w32_thread_list (dbg, pid, threads);
+	if (threads->length == 0) {
+		r_list_free (threads);
+		CloseHandle (ph);
+		return -1;
+	}
+	int tid = ((RDebugPid *)threads->head->data)->pid;
+	r_list_free (threads);
+	rio->ph = ph;
+	rio->debug = true;
+	return tid;
+}
+
+int w32_detach(RDebug *dbg, int pid) {
+	if (pid == -1) {
+		return false;
+	}
+	if (dbg->pid == pid) {
+		RIOW32Dbg *rio = dbg->user;
+		bool ret;
+		if (rio->debug) {
+			ret = DebugActiveProcessStop (pid);
+		}
+		CloseHandle (rio->ph);
+		rio->ph = NULL;
+		rio->debug = false;
+		return ret;
+	}
+	return false;
+}
+
+int w32_select(RDebug *dbg, int pid, int tid) {
+	RIOW32Dbg *rio = dbg->user;
+	if (rio->ph != (HANDLE)NULL) {
+		return true;
+	}
+	// hack to support w32dbg:// and attach://
+	DEBUG_EVENT de;
+	de.dwProcessId = pid;
+	de.dwThreadId = tid;
+	bool cont = true;
+	do {
+		if (!ContinueDebugEvent (de.dwProcessId, de.dwThreadId, DBG_CONTINUE)) {
+			break;
+		}
+		if (!WaitForDebugEvent (&de, 1000)) {
+			break;
+		}
+		switch (de.dwDebugEventCode) {
+		//case CREATE_PROCESS_DEBUG_EVENT:
+		//case CREATE_THREAD_DEBUG_EVENT:
+		case LOAD_DLL_DEBUG_EVENT:
+			{
+				HANDLE hf = de.u.LoadDll.hFile;
+				if (hf && hf != INVALID_HANDLE_VALUE) {
+					CloseHandle (hf);
+				}
+			} break;
+		case EXCEPTION_DEBUG_EVENT:
+			// TODO: check for the type of exception?
+			cont = false;
+			break;
+		default:
+			eprintf ("Unhandled debug event %d\n", de.dwDebugEventCode);
+			break;
+		}
+	} while (cont);
+	rio->ph = OpenProcess (PROCESS_ALL_ACCESS, FALSE, pid);
+	if (rio->ph == (HANDLE)NULL) {
+		return false;
+	}
+	rio->debug = true;
+	return true;
+}
+
+int w32_kill(RDebug *dbg, int pid, int tid, int sig) {
+	if (sig == 0) {
+		return true;
+	}
+	RIOW32Dbg *rio = dbg->user;
+	if (rio->debug) {
+		DebugActiveProcessStop (pid);
+	}
+	bool ret = false;
+	if (TerminateProcess (rio->ph, 1)) {
+		if (WaitForSingleObject (rio->ph, 1000) != WAIT_OBJECT_0) {
```

Comment:
this 1000 is used on every call of this function, what about setting it a global static const variable?

Reply:
That perhaps would be a good idea.